idaes/generic_models/unit_models/skeleton_model.py

```diff
@@ -0,0 +1,156 @@
+    with a generic skeleton to add a custom or a surrogate unit model
+    when controlvolumes are not required.
+    """
+
+    # This is a staticmethod that will be the default callable set for the
+    # initializer flag in the config block.
+    @staticmethod
+    def _default_initializer(model, opt=None, init_log=None, solve_log=None):
+        with idaeslog.solver_log(solve_log, idaeslog.DEBUG) as slc:
+                res = opt.solve(model, tee=slc.tee)
+        init_log.info("Initialization completed using default method {}.".
+                      format(idaeslog.condition(res)))
+
+    # Create Class ConfigBlock
+    CONFIG = ProcessBlockData.CONFIG()
+    CONFIG.declare("dynamic", ConfigValue(
+        default=False,
+        domain=In([False]),
+        description="Dynamic model flag",
+        doc="""Model does not support dynamics or holdup automatically.
+        The variables declared in this unit will need to be indexed with the
+        flowsheet time domain to facilitate connection with other
+        unit models."""))
+    CONFIG.declare("initializer", ConfigValue(
+        # Note: staticmethod is unbound, extract with __func__
+        default=_default_initializer.__func__,
+        description="Initializer to be used",
+        doc="""Flag to set the callback from user for initialization.
+        Default is set to use the default_initialize method."""))
+
+    def build(self):
+        """
+        General build method for SkeletonUnitModelBlockData.
+
+        Inheriting models should call `super().build`.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        super().build()
+
+
+    def add_ports(self, name, member_dict, doc=None):
+        """
+        This is a method to build Port objects in the SkeletonUnitModel and
+        populate them with appropriate port members as specified. User can add
+        as many inlet and outlet ports as required.
+
+        Keyword Args:
+            name : name to use for Port object.
+            dict : dictionary containing variables to be added to the port
+            doc : doc string for Port object
+
+        Returns:
+            A Pyomo Port object and associated components.
+        """
+        # Validate that member_dict is a dict
+        if not isinstance(member_dict, dict):
+            raise ConfigurationError(
+                "member_dict should be a dictionary "
+                "with the keys being the name assigned(strings) and "
+                "values being the variable objects declared in the model.")
+
+        # Create empty Port
+        p = Port(noruleinit=True, doc=doc)
+
+        # Add port object to model
+        setattr(self, name, p)
+
+        # Populate port and map names to actual variables as defined
+        for k in member_dict.keys():
+            p.add(member_dict[k], name=k)
+
+
+    def initialize(self, outlvl=idaeslog.NOTSET,
+                   solver=None, optarg=None):
```

Comment:
Should we add an optional argument for providing initial guesses? This would probably be passed directly to the initializer method, so users could do what they want with it.

Reply:
Interesting suggestion, however, how do you think we should handle that argument with the _default_initialize() method? Traditionally, we have assumed the initial guess to be applicable for the inlet port members. In the custom model case, the inlet port could be a subset of the input vars required for the model. I can always set `state_args` to `None` and not use it in the default_initialize and this could let the user use that argument in their custom method as they see fit. Let me know your thoughts.

Reply:
The simplest approach would to be accept it but then ignore it. Given that we don't know what a user will put inside a SkeletonModel, this is probably the best approach.

An alternative would be to assume these are initial values for variables in the outlet and try to assign to the values of these, but I am not sure that adds much (especially seeing as the main use for these models is for surrogates which are generally simple explicit functions).

Reply:
To add a little more - the main goal here is to give users some flexibility to pass extra information to their custom initializers. Obviously, this is not applicable to the default initializer, so it should not expect any additional information, but needs to have a placeholder for the addition information which it just ignores.

Reply:
Alright.